File: 15/activity15 Python/gdb/domain/salary_account.py
# gdb/domain/salary_account.py
from gdb.domain.account import Account
from gdb.domain.account_rules_engine import AccountRulesEngine

class SalaryAccount(Account):
    """Zero balance corporate salary account."""

    def __init__(self, account_number: int, name: str, age: int, initial_balance: float, tenure_years: int = 0) -> None:
        super().__init__(account_number, name, age, initial_balance, tenure_years)

    def get_account_type(self) -> str:
        return "Salary"

    def get_minimum_balance(self) -> float:
        return AccountRulesEngine.get_instance().get_minimum_balance("SALARY")

    def get_interest_rate(self) -> float:
        return AccountRulesEngine.get_instance().get_interest_rate("SALARY")

    def can_withdraw(self, amount: float) -> bool:
        return (self._balance - amount) >= self.get_minimum_balance()

File: activity15/src/com/gdb/domain/SalaryAccount.java
package com.gdb.domain;

import com.gdb.exceptions.*;

public class SalaryAccount extends AbstractAccount {
    private String employer;
    private int inactiveMonths;

    public SalaryAccount(String accountNumber, String name, int age, double balance, String status, String pin, String employer) {
        super(accountNumber, name, age, balance, "SALARY", status, pin, AccountRulesEngine.getSalaryDailyLimit());
        this.employer = employer;
        this.inactiveMonths = 0;
    }

    @Override
    public void processDebit(double amount) throws AccountException {
        if ((this.balance - amount) < 0) {
            throw new InsufficientBalanceException("Insufficient funds in salary account: available Rs " + balance);
        }
        this.balance -= amount;
    }

    public void incrementInactiveMonths() {
        this.inactiveMonths++;
        if (this.inactiveMonths >= 3) {
            this.status = "INACTIVE";
        }
    }

    public String getEmployer() { return employer; }
    public int getInactiveMonths() { return inactiveMonths; }
}
